colbert/index_updater.py

```diff
@@ -0,0 +1,433 @@
+import os
+import ujson
+import torch
+import numpy as np
+import tqdm
+
+from colbert.search.index_loader import IndexLoader
+from colbert.indexing.index_saver import IndexSaver
+from colbert.indexing.collection_encoder import CollectionEncoder
+
+from colbert.utils.utils import lengths2offsets, print_message, dotdict, flatten
+from colbert.indexing.codecs.residual import ResidualCodec
+from colbert.indexing.utils import optimize_ivf
+from colbert.search.strided_tensor import StridedTensor
+from colbert.modeling.checkpoint import Checkpoint
+from colbert.utils.utils import print_message, batch
+from colbert.data import Collection
+from colbert.indexing.codecs.residual_embeddings import ResidualEmbeddings
+from colbert.indexing.codecs.residual_embeddings_strided import ResidualEmbeddingsStrided
+from colbert.indexing.utils import optimize_ivf
+
+# For testing writing into new chunks, original default is 25000
+DEFAULT_CHUNKSIZE = 2 
+
+class IndexUpdater:
+    
+    '''
+    IndexUpdater takes in a searcher and adds/remove passages from the searcher.
+    A checkpoint for passage-encoding must be provided for adding passages.
+    IndexUpdater can also persist the change of passages to the index on disk.
+    
+    Sample usage:
+    
+        index_updater = IndexUpdater(config, searcher, checkpoint)
+        
+        added_pids = index_updater.add(passages) # all passages added to searcher with their pids returned
+        index_updater.remove(pids) # all pid within pids removed from searcher
+        
+        searcher.search() # the search now reflects the added & removed passages
+        
+        index_updater.persist_to_disk() # added & removed passages persisted to index on disk
+        searcher.Searcher(index, config) # if we reload the searcher now from disk index, the changes we made persists
+        
+    '''
+    
+    def __init__(self, config, searcher, checkpoint=None):
```

Comment:
Great! This design and the code are much cleaner than before.

Reply:
Thank you!